### PR TITLE
Add %X and %p formatters as well as LinkedList class

### PR DESCRIPTION
**Pull request by PeriBooty**
LinkedList
----
Eventually, when the `new` and `delete` operators are finished, the calloc and free calls should be replaced to use them.

### CODE REVIEW

File: spark/lib/lib.cpp [hunk 1]
```diff
@@ -174,9 +196,9 @@ char* itoa(int value, char* result, int base) {
     return rc;
 }
 
-void htoa(int64_t n, char* str) {
+void htoa(uint64_t n, char* str, int caps) {
```
Comment (VisualDevelopment) [inline spark/lib/lib.cpp]: Use a boolean here

File: spark/lib/lib.cpp [hunk 2]
```diff
@@ -174,9 +196,9 @@ char* itoa(int value, char* result, int base) {
...
     *str++ = '0';
-    *str++ = 'x';
+    *str++ = caps ? 'X' : 'x';
```
Comment (VisualDevelopment) [inline spark/lib/lib.cpp]: Making hex caps isn't making the 'x' caps too, only the preceding letters

File: spark/lib/linked_list.hpp [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    struct NodeLink* next;
+    int index;
+};
+
+template <class T>
+class LinkedList {
+    using Type = T;
+    using Link = NodeLink<T>;
+public:
+    LinkedList() : list(nullptr) {}
```
Comment (VisualDevelopment) [inline spark/lib/linked_list.hpp]: What is this supposed to do?

File: spark/lib/lib.cpp [hunk 4]
```diff
@@ -94,6 +95,27 @@ int sprintf(char* text, const char* format, ...) {
             for (size_t i = 0; i < len; i++)
                 text[i] = str[i];
 
+            text += len;
+            written += len;
+        } else if (*format == 'p') {
```
Comment (VisualDevelopment) [inline spark/lib/lib.cpp]: Also add %P

File: spark/lib/lib.hpp [hunk 5]
```diff
@@ -54,8 +54,9 @@ char* itoa(int value, char* result, int base);
  * 
  * @param n 
  * @param str 
+ * @param caps
  */
-void htoa(int64_t n, char* str);
+void htoa(uint64_t n, char* str, int caps);
```
Comment (VisualDevelopment) [inline spark/lib/lib.hpp]:
```diff
- void htoa(uint64_t n, char* str, int caps);
+ void htoa(uint64_t n, char* str, bool caps);
```

File: spark/lib/lib.cpp (hunk 4 above)
Comment (PeriBooty) [reply spark/lib/lib.cpp]: %P doesn't exist in C or C++, I'll still add it if you confirm that's intentional.

**Pull request merged**